ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -271,4 +489,376 @@ private void updateMapHelper(Map<String, Long> map, String key, Long value) {
+   * @param indexEntries a {@link List} of valid {@link IndexEntry} to be processed
+   * @param deletedKeys a {@link Map} of deleted keys to their corresponding deletion time in ms
+   */
+  private void processEntriesForLogSegmentBucket(ScanResults results, List<IndexEntry> indexEntries,
+      Map<StoreKey, Long> deletedKeys) {
+    for (IndexEntry indexEntry : indexEntries) {
+      IndexValue indexValue = indexEntry.getValue();
+      results.updateLogSegmentBaseBucket(indexValue.getOffset().getName(), indexValue.getSize());
+      if (!indexValue.isFlagSet(IndexValue.Flags.Delete_Index)) {
+        // put records
+        long expOrDelTimeInMs = indexValue.getExpiresAtMs();
+        if (deletedKeys.containsKey(indexEntry.getKey())) {
+          long deleteTimeInMs = deletedKeys.get(indexEntry.getKey());
+          expOrDelTimeInMs =
+              expOrDelTimeInMs != Utils.Infinite_Time && expOrDelTimeInMs < deleteTimeInMs ? expOrDelTimeInMs
+                  : deleteTimeInMs;
+        }
+        handleLogSegmentBucketUpdate(results, indexValue, expOrDelTimeInMs, SUBTRACT);
+      }
+    }
+  }
+
+  /**
+   * Determine if a given reference is within the boundary defined by the given start (inclusive) and end (exclusive)
+   * @param start the start of the bound (inclusive)
+   * @param end the end of the bound (exclusive)
+   * @param reference the reference to check against the boundary defined by start and end
+   * @return true if the reference is within the boundary, false otherwise
+   */
+  private boolean isWithinRange(long start, long end, long reference) {
+    return start <= reference && reference < end;
+  }
+
+  /**
+   * Runner that processes new entries buffered in the recentEntryQueue and make appropriate updates to keep the current
+   * {@link ScanResults} relevant.
+   */
+  private class QueueProcessor implements Runnable {
+    private volatile boolean cancelled = false;
+
+    @Override
+    public void run() {
+      try {
+        int entryCount;
+        ScanResults currentScanResults = scanResults.get();
+        // prevent new index entries getting dropped when IndexScanner is started before the QueueProcessor while
+        // new index entries are being added. e.g. IndexScanner started and recorded the end offset as 100, meanwhile
+        // new entries are added and IndexScanner is not finished yet. Before IndexScanner can finish, QueueProcessor
+        // is started and it's going to process new entries with offset 90 to 120. Since IndexScanner is still in
+        // progress the QueueProcessor will operate on the old ScanResults and index entries between 100 and
+        // 120 will be missing for the next forecast period.
+        if (isScanning || currentScanResults == null) {
+          return;
+        } else {
+          entryCount = queueEntryCount.get();
+        }
+        long processStartTimeMs = time.milliseconds();
+        for (int i = 0; i < entryCount; i++) {
```

Comment:
try to couple cancelled check with other conditional checks. for eg: 
 for (int i = 0; i < entryCount && !cancelled; i++) {